### PR TITLE
LearnerScorer: Support multiclass ranking

### PR DESCRIPTION
**Pull request by VesnaT**
score_data now returns numpy array, to support multiclass ranking (e.g. logistic regression coeficients for each class value). Scores for each class can be shown in Rank widget.
Similarly multiple PCA components can bw shown in Rank widget.

### CODE REVIEW

File: Orange/classification/logistic_regression.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ class _FeatureScorerMixin(LearnerScorer):
 
     def score(self, model):
         # Take the maximum attribute score across all classes
-        return np.max(np.abs(model.coefficients), axis=0)
+        return np.abs(model.coefficients)
```
Comment (kernc) [inline Orange/classification/logistic_regression.py]: Given the addendum, the comment is obsolete.

File: Orange/preprocess/score.py [hunk 2]
```diff
@@ -126,19 +126,32 @@ def score_data(self, data, feature=None):
...
-                      if attr in scores_grouped else 0
-                      for attr in data.domain.attributes]
+            return [sum(scores_grouped[attr]) / len(scores_grouped[attr])
+                    if attr in scores_grouped else 0
+                    for attr in data.domain.attributes]
 
-        return scores[data.domain.attributes.index(feature)] if feature \
-            else scores
+        if data.domain != self.domain:
+            if len(scores.shape) == 1:
```
Comment (kernc) [inline Orange/preprocess/score.py]: There's a `ndim` array property:
```py
assert arr.ndim == len(arr.shape)
```

File: Orange/preprocess/score.py [hunk 3]
```diff
@@ -126,19 +126,32 @@ def score_data(self, data, feature=None):
...
+                scores = average_scores(scores)
+            else:
+                empty_scores = np.empty((scores.shape[0],
+                                         len(data.domain.attributes)), float)
+                for i, row in enumerate(scores):
+                    empty_scores[i] = average_scores(row)
+                scores = empty_scores
```
Comment (kernc) [inline Orange/preprocess/score.py]: I think you can simplify do:
```py
else:
    scores = np.array([average_scores(row) for row in scores])
```

File: Orange/preprocess/score.py [hunk 4]
```diff
@@ -126,19 +126,32 @@ def score_data(self, data, feature=None):
...
+
+        if isinstance(scores, list):
+            scores = np.array(scores).reshape((1, len(scores)))
```
Comment (kernc) [inline Orange/preprocess/score.py]: ... and then move this up in the previous `if` branching.

File: Orange/tests/test_pca.py [hunk 5]
```diff
@@ -108,10 +108,27 @@ def test_chain(self):
...
+            scores = pca.score_data(data)
+            self.assertEqual(scores.shape,
+                             (pca.component, len(data.domain.attributes)))
+
+    def test_PCA_scorer_all_components(self):
+        data = Orange.data.Table('iris')
+        n_attr = len(data.domain.attributes)
+        pca = PCA()
+        scores = pca.score_data(data)
+        self.assertEqual(scores.shape, (n_attr, n_attr))
```
Comment (kernc) [inline Orange/tests/test_pca.py]: Please set _Settings > Editor > General > Ensure line feed at file end on Save_.

File: Orange/preprocess/score.py [hunk 6]
```diff
@@ -126,19 +126,28 @@ def score_data(self, data, feature=None):
...
+                    for attr in data.domain.attributes]
+
+        if data.domain != self.domain:
+            if scores.ndim == 1:
+                scores = average_scores(scores)
+            else:
+                scores = np.array([average_scores(row) for row in scores])
+
+        if isinstance(scores, list):
+            scores = np.array(scores).reshape((1, len(scores)))
```
Comment (kernc) [inline Orange/preprocess/score.py]: Nitpicking, sorry, but just to show how I might do it:
```py
if data.domain != self.domain:
    if scores.ndim == 1:
        scores = [scores]
    scores = np.array([average_scores(row) for row in scores])
else:
    scores = np.array([scores])
```

**Comment by acopar** [comment]
Logistic regression coefficients don't disappear if the connection between file widget and logistic regression widget is removed.

**Comment by kernc** [comment]
Was this closed intentionally and why?

**Pull request merged**